### ISSUE TITLE
Outputting text to the terminal increases run time by up to 30%

### ISSUE THREAD

**Issue by marcospb19** [ouch-org/ouch#77]
This percentage will vary depending on the terminal emulator you are using, I tried it with `alacritty`, which is GPU accelerated, and it also slowed it down by the same quantities (in some scenarios).

Differences were calculated by decompressing and compressing the `linux` arch package with and without redirecting the output to `/dev/null` with `"&> /dev/null"` at the end of the command.

This probably means that that we are flushing the output too many times, using a bufwriter will probably solve it.

However, there are some problems with regards to it, we need to make sure that `stderr`, `stdout` and `stdin` are always in sync. If we fail to sync `stdin`, we will wait for user input but the user won't see the message asking for it.

I'm actually not sure how difficult it is to solve this.

**Comment by marcospb19** [comment]
> How much increased time would be tolerable?

I'm not sure, we would need to investigate it further, we might even conclude at the end that our current time is tolerable or can be improved.

---

Let's see, I ran the same command and redirected the output to the file `output_file`.

```sh
ouch decompress --yes linux-5.14.9.arch2-1-x86_64.pkg.tar.zst &> output_file                                                                                                                                                                      
```

Now, we can show the contents in the terminal.

```sh
time cat output_file
```

It showed 0.145 seconds in the first run, and 0.045s for some of the consecutive runs.

That means that the terminal can display the same text in amount `0.045` seconds, if we take the minimum (this can be misleading if I'm missing something).

Here is the time it takes to decompressing the linux package with and without displaying the output:

```sh
time ouch decompress --yes linux-5.14.9.arch2-1-x86_64.pkg.tar.zst
# 0.95s on average
```

```sh
time ouch decompress --yes linux-5.14.9.arch2-1-x86_64.pkg.tar.zst &> output_file
# 0.82s on average
```

So it's an addition of 130 milliseconds.

**Comment by figsoda** [comment]
tested on https://github.com/ouch-org/ouch/commit/2c5a57c01ce8120ff55a148e0630fb4670bb7c49, got a slightly different result
```console
[...]
  Time (mean ± σ):      6.137 s ±  0.745 s    [User: 3.145 s, System: 2.873 s]
  Range (min … max):    5.589 s …  7.551 s    10 runs

Benchmark #2: target/release/ouch d linux.tar.gz -yo '/tmp/tmp.nhJZkRNvMD' > '/tmp/tmp.nhJZkRNvMD/log'
  Time (mean ± σ):      5.366 s ±  0.087 s    [User: 2.813 s, System: 2.406 s]
  Range (min … max):    5.242 s …  5.490 s    10 runs

Summary
  'target/release/ouch d linux.tar.gz -yo '/tmp/tmp.nhJZkRNvMD' > '/tmp/tmp.nhJZkRNvMD/log'' ran
    1.14 ± 0.14 times faster than 'target/release/ouch d linux.tar.gz -yo '/tmp/tmp.nhJZkRNvMD''
```

This is the script I tested with, requires hyperfine
```sh
#!/bin/sh

set -e

cargo build --release
[ ! -f linux.tar.gz ] && curl https://github.com/torvalds/linux/archive/v5.14.tar.gz -Lo linux.tar.gz
out=$(mktemp -d)
target/release/ouch d linux.tar.gz -o "$out"
hyperfine --show-output \
    "target/release/ouch d linux.tar.gz -yo '$out'" \
    "target/release/ouch d linux.tar.gz -yo '$out' > '$out/log'"
rm -r "$out"
```

**Comment by marcospb19** [comment]
should be fixed once #643 is solved.

**Comment by marcospb19** [comment]
Fixed! #643 was solved by #642 :tada:
